src/net/server/dgram.rs

```diff
@@ -0,0 +1,745 @@
+                write_timeout: _,     // TO DO
+                middleware_chain: _,  // TO DO
+            }) => {
+                // TODO: Support dynamic replacement of the middleware chain?
+                // E.g. via ArcSwapOption<MiddlewareChain> instead of Option?
+            }
+
+            ServerCommand::Shutdown => {
+                // Stop receiving new messages.
+                return Err("Shutdown command received".to_string());
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn recv_from(
+        &self,
+    ) -> Result<(Buf::Output, SocketAddr, usize), io::Error> {
+        let mut res = self.buf.create_buf();
+        let (addr, bytes_read) = {
+            let mut buf = ReadBuf::new(res.as_mut());
+            let addr = poll_fn(|ctx| self.sock.poll_recv_from(ctx, &mut buf))
+                .await?;
+            (addr, buf.filled().len())
+        };
+        Ok((res, addr, bytes_read))
+    }
+
+    async fn send_to(
+        sock: &Sock,
+        data: &[u8],
+        dest: &SocketAddr,
+        limit: Duration,
+    ) -> Result<(), io::Error> {
+        let send_res =
+            timeout(limit, poll_fn(|ctx| sock.poll_send_to(ctx, data, dest)))
+                .await;
+
+        let Ok(send_res) = send_res else {
+            return Err(io::ErrorKind::TimedOut.into());
+        };
+
+        let sent = send_res?;
+
+        if sent != data.len() {
+            Err(io::Error::new(io::ErrorKind::Other, "short send"))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn mk_state_for_request(
+        &self,
+    ) -> RequestState<Sock, Buf::Output, Svc::Target> {
+        RequestState::new(
+            self.sock.clone(),
+            self.command_tx.clone(),
+            self.config.write_timeout,
+        )
+    }
+}
+
+//--- MessageProcessor
+
+impl<Sock, Buf, Svc> MessageProcessor<Buf, Svc>
+    for DgramServer<Sock, Buf, Svc>
+where
+    Sock: AsyncDgramSock + Send + Sync + 'static,
+    Buf: BufSource + Send + Sync + 'static,
+    Buf::Output: Octets + Send + Sync + 'static,
+    Svc: Service<Buf::Output> + Send + Sync + 'static,
+{
+    type State = RequestState<Sock, Buf::Output, Svc::Target>;
+
+    fn add_context_to_request(
+        &self,
+        request: Message<Buf::Output>,
+        received_at: Instant,
+        addr: SocketAddr,
+    ) -> Request<Message<Buf::Output>> {
+        let ctx =
+            TransportSpecificContext::Udp(UdpSpecificTransportContext {
+                max_response_size_hint: self.config.max_response_size,
+            });
+        Request::new(addr, received_at, request, ctx)
+    }
+
+    fn process_call_result(
+        call_result: CallResult<Buf::Output, Svc::Target>,
+        addr: SocketAddr,
+        state: RequestState<Sock, Buf::Output, Svc::Target>,
+        metrics: Arc<ServerMetrics>,
+    ) {
+        metrics.num_pending_writes.fetch_add(1, Ordering::Relaxed);
+
+        tokio::spawn(async move {
```

Comment:
Firstly, a related fragment from the [Tokio docs](https://tokio.rs/tokio/tutorial/spawning#tasks):

> Tasks in Tokio are very lightweight. Under the hood, they require only a single allocation and 64 bytes of memory. Applications should feel free to spawn thousands, if not millions of tasks.

This spawn happens in a sync function which was called inside a task spawned by `MessageProcessor::postprocess_response()`, so it's also notable that we spawn a task having just spawned a task. This double spawning is a consequence of a fight I had with the Rust compiler that I didn't resolve about trying to make the callback in the service be an async function, so for now it remains a sync function. I'd like to resolve this, and even revisit the whole request pre/post processing and middleware chain to make it more Hyper stackable service like, but that isn't possible now in the time available.

So, we need to do the write from inside an async context because the `AsyncDgramSock::poll_send_to()` fn is intended for use in an async context, and because we wrap it inside an async timeout, and so we spawn it in a task to do that... and the Tokio docs say task spawning is cheap, and also I expect the actual application logic and DNS message processing to be way more expensive than the overhead of spawning a single Tokio task.